### ISSUE TITLE
Next app remounts on navigating between pages with serverSideTranslations to w/o it and vice versa

### ISSUE THREAD

**Issue by vc-vishal** [isaachinman/next-i18next#1769]
## Bug Description 
When we navigate between page with serverSideTranslations to page without serverSideTranslations or vice versa, whole next app remounts. 

## Occurs in next-i18next version
10.5.0

## Steps to reproduce
[https://github.com/vc-vishal/next-i18next-localization](url)

## Expected behaviour
If possible then app should not remount in this case.

## OS
 - Browser: Crome Version 100.0.4896.60 (64-bit)

## Additional context
I my app, I have to play music in 4-5 pages. But after language selection whole app remounts and music start from begin.
Please let me know if any mistake found in demo or in my understanding.

**Comment by isaachinman** [comment]
Can you debug why exactly that is occurring? It's most likely [this](https://github.com/isaachinman/next-i18next/blob/master/src/appWithTranslation.tsx#L74) ternary, right?

Can you put your music functionality in a HOC itself, that wraps `appWithTranslation`?

**Comment by aNyMoRe0505** [comment]
@isaachinman  hello
if we change language via [next/router](https://nextjs.org/docs/advanced-features/i18n-routing#transition-between-locales) also cause remount. Is it intentional ?

**Comment by isaachinman** [comment]
@aNyMoRe0505 Yes, you can see the logic [here](https://github.com/isaachinman/next-i18next/blob/master/src/appWithTranslation.tsx#L79).

**Comment by vc-vishal** [comment]
@isaachinman If I put music functionality in HOC (i.e. MyApp) then also my problem will not solve. 

The problem is language change remounts the whole app. That will restarts the music from begin. And my requirement is to continue play music between some pages.

**Comment by vc-vishal** [comment]
> Can you debug why exactly that is occurring? It's most likely [this](https://github.com/isaachinman/next-i18next/blob/master/src/appWithTranslation.tsx#L74) ternary, right?
> 
> Can you put your music functionality in a HOC itself, that wraps `appWithTranslation`?

@isaachinman If I put music functionality in HOC (i.e. MyApp) then also my problem will not solve.

The problem is language change remounts the whole app. That will restarts the music from begin. And my requirement is to continue play music between some pages.

**Comment by isaachinman** [comment]
@vc-vishal Yes, `appWithTranslation` remounts on language change. My suggestion for a HOC was to put it one level above `appWithTranslation`.

**Comment by vc-vishal** [comment]
@isaachinman I got your point but I've already wrapped `_app.js` app component with `appWithTranslation`. I have not any levels above that.

If you have any solution regarding this then tell me.

**Comment by rtrembecky** [comment]
> @vc-vishal Yes, appWithTranslation remounts on language change. My suggestion for a HOC was to put it one level above appWithTranslation.

@isaachinman, why is this the case? I think in order to finally close this issue (I consider it a successor of #1075 and #1710), something about this should at least be added to README if there's really no way around this issue.

### Issue summary and the current state
As a user, with the current state of documentation, I expect to be able to wrap the whole next.js app with `appWithTranslation` and be done with changes in `_app.tsx`, keeping all of the staying app's functionality while enjoying the benefits of the localization context. However, as we see, the app remounts on language change, losing its state (e.g. in context providers). Also, the solution is non-trivial.

### Suggestions for improvement
If this is a feature:
- README adjustments - mention this feature and add an example code snippet (e.g. as in my previous comment)
- what about a new helper for this scenario? Currently, what I meant by the solution being "non-trivial", is the need for splitting the single `App` component definition, creating the `App`, `TranslatedApp`, and `EnhancedApp` components (using my naming above). Maybe we could export such `TranslatedApp` from the package? Potentially giving it an optional `wrapper` prop to take a HOC to wrap the `<Component {...pageProps} />` JSX (I don't use this above, but I can imagine a use case where you have some top-level HOCs that depend on the localization context). Then again, the documentation would probably be enough.

If not really, I hope we can find another solution.

In any way, I'd like to understand more about this. Is there some technical issue why the app remount is the only way and there's no way to let React reconciliation know the components are still the same and shouldn't remount (losing state)?

**Comment by rtrembecky** [comment]
By the way, I also created a reproducible example with a single page: https://github.com/rtrembecky/next-i18next-app-remount-repro

Here is a preview:

https://user-images.githubusercontent.com/6044119/166961043-f76a66e0-082e-4453-b041-397bd20942ab.mp4

**Comment by vc-vishal** [comment]
> @vc-vishal the suggested solution is this structure:
> 
> ```ts
> const App: FC<AppProps> = ({ Component, pageProps }) => <Component {...pageProps} />
> 
> // doesn't wrap whole app because it causes remounts on route change - probably a bug: https://github.com/isaachinman/next-i18next/issues/1075
> const TranslatedApp = appWithTranslation(App)
> 
> const EnhancedApp: FC<AppProps> = (props) => {
>   return (
>     <MusicFunctionalityProvider>
>       <TranslatedApp {...props} />
>     </MusicFunctionalityProvider>
>   )
> }
> 
> export default EnhancedApp
> ```
> 
> it works in my app to avoid resetting the context in providers wrapping the app. if you can lift the state up to such `MusicFunctionalityProvider`, you won't lose it on locale change or navigation.

@rtrembecky  That worked for me. Thank you for your solution and demo. ❤

**Comment by rtrembecky** [comment]
> @rtrembecky  That worked for me. Thank you for your solution and demo. ❤

@vc-vishal You're welcome 🙂 I hope more people will find it helpful.

**Comment by isaachinman** [comment]
As per https://github.com/isaachinman/next-i18next/discussions/1796, it may be worthwhile for anyone interested in this issue to experiment with removing this line:

https://github.com/isaachinman/next-i18next/blob/33a184e609daabe6c15a0cfbc879120b3e44f0ee/src/appWithTranslation.tsx#L79

Historically it was needed to force rerenders on locale change, but it may no longer be necessary.
